gdal/gcore/gdal_priv.h

```diff
@@ -659,20 +660,32 @@ class CPL_DLL GDALDataset : public GDALMajorObject
                 struct Private;
                 std::unique_ptr<Private> m_poPrivate;
             public:
+
+                using value_type = OGRLayer*;
+                using reference = OGRLayer*;
+                using difference_type = void;
+                using pointer = void;
+                using iterator_category = std::input_iterator_tag;
```

Comment:
Are these `using`s actually used?

Reply:
These are required in order for the iterator to be a standard InputIterator. The advantage of conforming to the standard concept is that then you can use the iterators / range with algorithms that require it. And yes, algorithms do use these.  For instance, a simple zip iterator would be quite likely to require InputIterators as its input.

Reply:
Then a comment explaining would be good as I don't seem them referenced and it will help others understand how to do this.

Reply:
There is no problem adding a comment, but do keep in mind that these typedefs are required for *any* standard conforming iterator class. So even if they look orphaned to you now, you will soon get used to it. 

How about the following:

```cpp
// typedefs required by Iterator concept:
using value_type =... etc.
```

Reply:
I'd appreciate if autotest/cpp/test_ogr.cpp could be enriched with some test that would use some std algorithm that would check that the iterator really implements what is needed.

Reply:
std::copy is an example that requires InputIterator, The test could copy the layers into a std::vector<OGRLayer*> and then compare that all the pointers in the GDALDataset and std::vector are the same. I am sorry but cannot create that test though.